src/ray/core_worker/actor_creator.h

```diff
@@ -67,20 +67,20 @@ class DefaultActorCreator : public ActorCreatorInterface {
   explicit DefaultActorCreator(std::shared_ptr<gcs::GcsClient> gcs_client)
       : gcs_client_(std::move(gcs_client)) {}
 
-  Status RegisterActor(const TaskSpecification &task_spec) override {
-    auto promise = std::make_shared<std::promise<void>>();
-    auto status = gcs_client_->Actors().AsyncRegisterActor(
-        task_spec, [promise](const Status &status) { promise->set_value(); });
-    if (status.ok() &&
-        promise->get_future().wait_for(std::chrono::seconds(
+  Status RegisterActor(const TaskSpecification &task_spec) const override {
+    auto promise = std::make_shared<std::promise<Status>>();
+    RAY_UNUSED(gcs_client_->Actors().AsyncRegisterActor(
+        task_spec, [promise](const Status &status) { promise->set_value(status); }));
+    auto future = promise->get_future();
+    if (future.wait_for(std::chrono::seconds(
             ::RayConfig::instance().gcs_server_request_timeout_seconds())) !=
-            std::future_status::ready) {
+        std::future_status::ready) {
       std::ostringstream stream;
       stream << "There was timeout in registering an actor. It is probably "
                 "because GCS server is dead or there's a high load there.";
       return Status::TimedOut(stream.str());
     }
-    return status;
+    return future.get();
```

Comment:
The bug was that `AsyncRegisterActor ` always returns the Status::OK() although the RegisterActor RPC failed (which we should retrieve from promise instead). I realized this was already properly used for placement group RPCs, but not for this particular actor RPC.

Reply:
Since Registering named actor is synchronous, it is guaranteed that there's no race condition here. I will write a comment.